### ISSUE TITLE
Use_awaitable is incompatible to the "Asynchronous Operations" doc.

### ISSUE THREAD

**Issue by joka921** [chriskohlhoff/asio#969]
According to the documentation for the requirements on asynchronous operations,
[(](https://www.boost.org/doc/libs/1_78_0/doc/html/boost_asio/reference/asynchronous_operations.html)
there are two ways to specify a custom async operation:
- manually via `async_result<decay_t<CompletionToken>, Signature>`
 ( First manually obtain a `completion_handler` from the `CompletionToken`, then instantiate
 an `async_result` using this token.
- using the convenience wrapper `async_completion<CompletionToken, Signature>
  (no need for `decay_t`, directly gives you the `async_result` and the `completion_handler`.

However, aync operations implemented with these mechanisms do not work, when the completion token
is `use_awaitable`. (`use_awaitable_t<>` does not define a `completion_handler_type`, but only a `handler_type`, and this `handler_type` can't be constructed from `use_awaitable`, but always requires an additional Executors.
These operations  work fine with other completion tokens like `use_future`.

After discovering this, I was a bit puzzled since `built-in` async operations work with `using_awaitable`.
I looked into their implementation, and they use `async_initiate<CompletionToken, Signature>`.
Using this pattern, I could make my asynchronous functions work, but this method has the serious drawback, that `async_initiate` has no documentation whatsoever.

My general issue is that the documentation and the reality being out of sync.
I would suspect, that the issue lies with the implementation of `use_awaitable_t`.
If this can be confirmed I could look into the implementation `async_initiate` to see, how this can be fixed,
but I first want to have some confirmation, that this is a new issue, and that there is no deep underlying problem, of which I am not aware.

In the following, there is a small program, that demonstrates the issue, it is also available on godbolt:
[](https://godbolt.org/z/z1n9q3rds)

I am sorry for the lengthy issue, but it also contains quite a story:)
Best regards and happy holidays
Johannes

```
#include <boost/asio.hpp>

using R1 = std::exception_ptr;
using R2 = int;
using std::decay_t;
using namespace boost::asio;


template<class CompletionToken>
auto async_manually(CompletionToken&& token)
{
  typename async_result<decay_t<CompletionToken>, void(R1, R2)>::completion_handler_type
    completion_handler(forward<CompletionToken>(token));

  async_result<decay_t<CompletionToken>, void(R1, R2)> result(completion_handler);


  // initiate the operation and cause completion_handler to be invoked with
  // the result
  completion_handler(nullptr, 1);


  return result.get();
}


template<class CompletionToken>
auto async_simple(CompletionToken&& token)
{
  async_completion<CompletionToken, void(R1, R2)> init(token);

  // initiate the operation and cause init.completion_handler to be invoked
  // with the result
  init.completion_handler(nullptr, 1);

  return init.result.get();
}


template<class CompletionToken>
auto async_as_in_library(CompletionToken&& token)
{

  // initiate the operation and cause init.completion_handler to be invoked
  // with the result
  auto initiating = [](auto&& handler) {
      handler(nullptr, 1);
  };

  return async_initiate<CompletionToken, void(R1, R2)>(initiating, std::forward<CompletionToken>(token));
}

awaitable<int> with_awaitable() {
    // A async_manually and async_simple do not compile with use_awaitable
    //auto a = co_await async_manually(use_awaitable);
    auto fa = async_manually(use_future);
    //auto b = co_await async_simple(use_awaitable);
    auto fb = async_simple(use_future);
    auto c = co_await async_as_in_library(use_awaitable);
    auto fc = async_as_in_library(use_future);
;}



int main(void) {
 
}
```

**Comment by Maximsiv1410** [comment]
Hi
Why do you try to invoke your completion token like this? 
`completion_handler(nullptr, 1);`
Since you have only one argument, try it this way:
```
template<class CompletionToken>
auto async_simple(CompletionToken&& token)
{
  async_completion<CompletionToken, void(int)> init(token);

  // initiate the operation and cause init.completion_handler to be invoked
  // with the result
  init.completion_handler(/* some int value */);

  return init.result.get();
}
```
But it would be much more natural to call completion token outiside of this function

**Comment by Maximsiv1410** [comment]
Hello, could you explain why do you use exception ptr here?

**Comment by Maximsiv1410** [comment]
So I looked through source code of asio, there is another way to use async_result
asio::use_awaitable has `initiate `member function, which does some additional stuff 
You can see it here:
```
template <typename CompletionToken,
    ASIO_COMPLETION_SIGNATURE... Signatures,
    typename Initiation, typename... Args>
inline typename constraint<
    detail::async_result_has_initiate_memfn<
      CompletionToken, Signatures...>::value,
    ASIO_INITFN_DEDUCED_RESULT_TYPE(CompletionToken, Signatures...,
      (async_result<typename decay<CompletionToken>::type,
        Signatures...>::initiate(declval<ASIO_MOVE_ARG(Initiation)>(),
          declval<ASIO_MOVE_ARG(CompletionToken)>(),
          declval<ASIO_MOVE_ARG(Args)>()...)))>::type
async_initiate(ASIO_MOVE_ARG(Initiation) initiation,
    ASIO_NONDEDUCED_MOVE_ARG(CompletionToken) token,
    ASIO_MOVE_ARG(Args)... args)
{
  return async_result<typename decay<CompletionToken>::type,
    Signatures...>::initiate(ASIO_MOVE_CAST(Initiation)(initiation),
      ASIO_MOVE_CAST(CompletionToken)(token),
      ASIO_MOVE_CAST(Args)(args)...);
}
```
This overload of async_initiate is chosen at compile time via this `async_result_has_initiate_memfn` (SFINAE I guess)
So you should use this part manually
```
 return async_result<typename decay<CompletionToken>::type,
    Signatures...>::initiate(ASIO_MOVE_CAST(Initiation)(initiation),
      ASIO_MOVE_CAST(CompletionToken)(token),
      ASIO_MOVE_CAST(Args)(args)...);
```
or just use async_initiate since it was created in order to hide underlying complexity of library
I agree - there is lack of documentation targeted at async_initiate or what's more 'async_compose' which suites more for stateful ops
But there are some useful [examples ](https://github.com/chriskohlhoff/asio/tree/master/asio/src/examples/cpp11/operations)which can help you

**Comment by joka921** [comment]
Hi Max,
Thanks for your comments and your research into the details of use_awaitable.
1. In my examples I use exception_ptr, because in my real-life scenario, the async operation might throw an arbitrary exception,
    And to my understanding, the completion signature   void(exception_ptr, resultType)  is expected in this case (at least it works for the builtin completion tokens use_future and use_awaitable, as you see in my examples.

2. I also looked into why "async_initiate" works in contrast to the other two methods and also found the initiate member function.
    Without having understood every detail it seems to me, that use_awaitable requires this additional magic, which cannot easily
    be put into the "old" interface of directly obtaining a completion handler manually.

3. So probably (@chriskohlhoff) the real issue is, that you should async_initiate to define operations if you want them to be compatible with all types of Completion tokens, especially with use_awaitable and that this information is not contained in the corresponding documentation. If someone who is more into the details of use_awaitable could confirm this, I will happily reformulate this issue.

**Comment by Maximsiv1410** [comment]
Well, I went further and here's what I've dig up:
If you want your async_ operation to throw exceptios, you need this:

```
template<typename CompletionToken>
auto async_with_exception(asio::io_context& ctx, CompletionToken&& token) {
	return asio::async_initiate<CompletionToken, void(std::exception_ptr, int)>
		([&](auto&& token) {
		asio::post(ctx,
		[token = std::forward<decltype(token)>(token)]() mutable
		{
			try {
				throw std::runtime_error("emulated exception");
			}
			catch (...) {
				token(std::current_exception(), 3);
			}
		});
	}, token);
}

asio::awaitable<void> run_exception(asio::io_context& ctx) {
	try {
		auto value = co_await async_with_exception(ctx, asio::use_awaitable);
		std::cout << "run_exception: " << value << std::endl;
	}
	catch (std::runtime_error const& ex) {
		std::cerr << "run_exception: " << ex.what() << std::endl;
	}	
}

int main() {
	asio::io_context ctx;
	asio::co_spawn(ctx, run_exception(ctx), asio::detached);
	ctx.run();
}
```
However if you want to have old-style error_code handling, here you are:
```
template<typename CompletionToken>
auto async_with_error_code(asio::io_context & ctx, CompletionToken&& token) {
	return asio::async_initiate<CompletionToken, void(std::error_code, int)>
	([&](auto && token)	{
		asio::post(ctx, 
		[token = std::forward<decltype(token)>(token)]() mutable 
		{			
			token(asio::error::access_denied, 2);
		});
	}, token);
}

asio::awaitable<void> run_tuple_error(asio::io_context& ctx) {
	auto [ec, value] = co_await async_with_error_code(ctx, asio::experimental::as_tuple(asio::use_awaitable));
	if (ec) {
		std::cout << "run_tuple_error: " << ec.message() << std::endl;
	}
	else {
		std::cout << "run_tuple_error: " << value << std::endl;
	}
}

int main() {
	asio::io_context ctx;
	asio::co_spawn(ctx, run_tuple_error(ctx), asio::detached);
	ctx.run();
}
```
Note:
you can drop 'as_tuple' here and then you get your error_code thrown as asio::system_error
```
asio::awaitable<void> run_throw_error(asio::io_context& ctx) {
	try {
		auto value = co_await async_with_error_code(ctx, asio::use_awaitable);
		std::cout << "run_throw_error: " << value << std::endl;
	}
	catch (asio::system_error const & ex) {
		std::cout << "run_throw_error: " << ex.what() << std::endl;
	}	
}
```

And also the simplest silent way without any error handling:
```
asio::awaitable<void> run_silent(asio::io_context & ctx) {	
	auto value = co_await async_silent(ctx, asio::use_awaitable);
	std::cout << "run_silent: " << value << std::endl;
}
```

**Comment by joka921** [comment]
@Maximsiv1410 
Thanks for your efforts. I was aware of how to handle exceptions,
I just omitted them in my example, because the exception handling was orthogonal to my issue.
(I just used the `void(exception_ptr, int)` syntax as a minimal
signature that works with `use_awaitable` and supports values as well as errors.

For your information (not really related to the original issue).
My actual use case was quite similar to your example,
but I wanted to run the actual code outside of the io_context in a new thread, so I basically ended up with your code,
but with `std::jthread(<lambda-that-uses-completion-handler>).detach()` instead of `asio::post(<same-lambda>)`

**Comment by Tradias** [comment]
The revision history explains that the old style of `async_result` has been superseded by the `initiate` member function since 1.70.0 https://www.boost.org/doc/libs/1_78_0/doc/html/boost_asio/history.html#boost_asio.history.asio_1_14_0___boost_1_70

It also explicitly says that:

> Please note that the use_awaitable and redirect_error completion tokens work only with asynchronous operations that use the new form of async_result with member function initiate.

Like Maximsiv1410 said, prefer the use of `async_initiate` instead of `async_result<>::initiate` to avoid having to deal with these implementation details.
